### DISCUSSION TITLE
"global" variables i1,i2,v2,v2, ...

### DISCUSSION THREAD

**Discussion by Peter-Jacob** [adesutherland/CREXX#271] [Q&A]
As the new IRAND (random) instruction doesn't compile anymore, I noticed that all "global" variables as i1,i2,v1,v2, ... have been dropped and need to be defined locally in each instruction that needs it. To avoid duplicated definitions, the instruction needs to be defined as a block {}. Isn't that creating overhead and a larger program size?

**Comment by adesutherland** [comment]
I made the change to get a small performance improvement (unproven by the way!) and for maintainability.

The problem with the old approach was that the C compiler would never be able to work out if the write to, for example, i2 could be optimised out because so far as it was concerned i2 could be critically needed for the next time round the interpreter loop. Well actually a clever compiler might/should be able to do this ... but I didn't believe it!

By having it in its own { } block the compiler knows the [very limited] scope of the variable and (for example) might decide to use a register for it (or optimise it out altogether). Untested.

I don't believe the program will be bigger because of the optimiser but this is also untested! That said - I forced more inlining which will make it bigger for sure.

I think it is easier to maintain because the scope / type / purpose of the variables is clearer.

That was my logic!

**Comment by Peter-Jacob** [comment]
ok, that makes sense! No problem, I changed my code anyway.
